_posts/2021-03-05-automatically-generated-api-tests-with-clojure-and-reitit.md

```diff
@@ -0,0 +1,332 @@
+---
+layout: post
+title: Automatically generated API tests with Clojure and Reitit
+author: jgke
+excerpt: >
+  Is your entire API tested? Maybe you should test that.
+tags:
+ - Clojure
+ - Reitit
+ - Testing
+---
+
+Have you ever forgotten to write a test for some API endpoint, which then
+eventually happens to break? Would it be possible to check whether all of your
+API is tested? If you are using a data-driven router, it's possible to write a
+test which walks over the entire API.
+
+My current customer project is using something along these lines to test that
+the entire API handles credentials and permissions appropriately and doesn't
+randomly crash at some endpoints. As a consequence, the tests generated by
+these methods happen to create an almost complete smoke test suite, where a new
+endpoint can be added with a few lines of code. It also forces our API to behave
+consistently and give out similar errors independent of the endpoint.
+
+Example API
+===========
+
+For a small example api, we have the following endpoints:
+- `POST:/api/resource` - Create a new resource
+- `GET:/api/resource/:resource-id` - Get the created resource by ID
+- `PUT:/api/resource/:resource-id` - Update the resource
+- `POST:/api/authorization` - Add a user to a resource
+- `DELETE:/api/authorization` - Delete a user from a resource
+
+Only users added to a resource can update it. The creator of the resource is
+added when creating the resource.
+
+The endpoints behave in the following way: All requests without the Auth header
+return an error. Creating resources returns the created resource id. The id can
+be used to find the created resource.
+
+```
+$ curl -d "Hello" localhost:3001/api/resource   
+{:cause :missing-authentication}
+$ curl -H "Auth: 1" -H "Content-Type: application/json" -d '{"content": "Hello"}' localhost:3001/api/resource
+{:resource-id "4"}
+$ curl -H "Auth: 1" localhost:3001/api/resource/4
+{:content "Hello"}
+```
+
+Using a different Auth id than the one used to create the resource initially
+returns an error, but if the original user is used to add the other user into
```

Comment:
"…but if the original user adds a new user into the resource, the new user can also access it." might be slightly more succinct.